velox/substrait/VeloxToSubstraitPlan.cpp

```diff
@@ -170,13 +176,119 @@ void VeloxToSubstraitPlanConvertor::toSubstrait(
+      1, sources.size(), "Aggregation plan node must have exactly one source.");
+  const auto& source = sources[0];
+
+  // Build source.
+  toSubstrait(arena, source, aggregateRel->mutable_input());
+
+  // Convert aggregate grouping keys, such as: group by key1, key2.
+  auto inputType = source->outputType();
+  auto groupingKeys = aggregateNode->groupingKeys();
+  int64_t groupingKeySize = groupingKeys.size();
+  ::substrait::AggregateRel_Grouping* aggGroupings =
+      aggregateRel->add_groupings();
+
+  for (int64_t i = 0; i < groupingKeySize; i++) {
+    aggGroupings->add_grouping_expressions()->MergeFrom(
+        exprConvertor_->toSubstraitExpr(
+            arena,
+            std::dynamic_pointer_cast<const core::ITypedExpr>(
+                groupingKeys.at(i)),
+            inputType));
+  }
+
+  // AggregatesSize should be equal to or greater than the aggregateMasks Size.
+  // Two cases: 1. aggregateMasksSize = 0, aggregatesSize > aggregateMasksSize.
+  // 2. aggregateMasksSize != 0, aggregatesSize = aggregateMasksSize.
+  auto aggregates = aggregateNode->aggregates();
+  auto aggregateMasks = aggregateNode->aggregateMasks();
+  int64_t aggregatesSize = aggregates.size();
+  int64_t aggregateMasksSize = aggregateMasks.size();
+  VELOX_CHECK_GE(aggregatesSize, aggregateMasksSize);
+
+  for (int64_t i = 0; i < aggregatesSize; i++) {
+    ::substrait::AggregateRel_Measure* aggMeasures =
+        aggregateRel->add_measures();
+
+    auto aggMaskExpr = aggregateMasks.at(i);
+    // Set substrait filter.
+    ::substrait::Expression* aggFilter = aggMeasures->mutable_filter();
+    if (aggMaskExpr.get()) {
+      aggFilter->MergeFrom(exprConvertor_->toSubstraitExpr(
+          arena,
+          std::dynamic_pointer_cast<const core::ITypedExpr>(aggMaskExpr),
+          inputType));
+    } else {
+      // Set null.
+      aggFilter = NULL;
+    }
+
+    // Process measure, eg:sum(a).
+    const auto& aggregatesExpr = aggregates.at(i);
+    ::substrait::AggregateFunction* aggFunction =
+        aggMeasures->mutable_measure();
+
+    // Aggregation function name.
+    const auto& funName = aggregatesExpr->name();
+    // set aggFunction args.
+    const auto& aggregatesExprInputs = aggregatesExpr->inputs();
+
+    int64_t aggregatesExprInputsSize = aggregatesExprInputs.size();
+    for (int64_t j = 0; j < aggregatesExprInputsSize; j++) {
+      ::substrait::Expression* aggFunctionExpr = aggFunction->add_args();
+      core::TypedExprPtr expr = aggregatesExprInputs[j];
+
+      // If the expr is CallTypedExpr, people need do project firstly.
+      if (auto aggregatesExprInput =
+              std::dynamic_pointer_cast<const core::CallTypedExpr>(expr)) {
+        VELOX_NYI("In Velox Plan, the aggregates type cannot be CallTypedExpr");
+      } else if (
+          auto aggregatesExprInput =
+              std::dynamic_pointer_cast<const core::ConstantTypedExpr>(expr) &&
+              funName == "count") {
+        // For case count(1).
```

Comment:
This looks like a hack. It is valid for other functions to also have constant inputs. Why not process these in a generic way?

Reply:
This is because we need to use ` count(1)` to represent `count(*)`. In the substrait, the representation of  `count(*)` function is different from that of other functions with constant parameters. `count(*)` needs to be expressed as ` "name": "count:opt"` , and the `args` in the substrait plan need to be empty; `count(col)` needs to be expressed as: `"name": "count:opt_any"`, while other functions with constant inputs, for example, `sum(1)`, it needs to be expressed as:` "name": "sum:opt_i32"` and the`args` in plan is not empty.

Reply:
What is "opt", "opt_any" and "opt_i32" here?

Reply:
opt is `Optional Enumeration`, see [here](https://substrait.io/extensions/#function-signature-compound-names:~:text=are%20described%20below.-,Function%20Signature%20Compound%20Names,%C2%B6,-A%20YAML%20file)

Reply:
i32 is the type of inputs. any is any type.

Reply:
Thank you for the reference. This is helpful. Do you also have a pointer to where the special rules for "count" are specified?

Reply:
this was found by debug the code of omnisci.

Reply:
> this was found by debuging the code of omnisci.

In this case, can this be not a requirement from Substrait, but rather some requirement from the implementation in omnisci? Is this a useful one-off or should we rather fix it there?

CC: @jacques-n @cpcloud

Reply:
Remove the special rules for count(*) , we will handle it internally.